warehouse/email/ses/views.py

```diff
@@ -0,0 +1,125 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+from pyramid.httpexceptions import HTTPBadRequest
+from pyramid.response import Response
+from pyramid.view import view_config
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.sql import exists
+
+from warehouse.email.ses.models import EmailMessage, EmailStatus, Event
+from warehouse.utils import sns
+
+
+def _verify_sns_message(request, message):
+    verifier = sns.MessageVerifier(
+        topics=[request.registry.settings["mail.topic"]],
+        session=request.http,
+    )
+
+    try:
+        verifier.verify(message)
+    except sns.InvalidMessage as exc:
+        raise HTTPBadRequest(str(exc)) from None
+
+
+@view_config(
+    route_name="ses.hook",
+    require_methods=["POST"],
+    require_csrf=False,
+    header="x-amz-sns-message-type:SubscriptionConfirmation",
+)
+def confirm_subscription(request):
+    data = request.json_body
+
+    # Check to ensure that the Type is what we expect.
+    if data.get("Type") != "SubscriptionConfirmation":
+        raise HTTPBadRequest("Expected Type of SubscriptionConfirmation")
+
+    _verify_sns_message(request, data)
+
+    aws_session = request.find_service(name="aws.session")
+    sns_client = aws_session.client(
+        "sns",
+        region_name=request.registry.settings.get("mail.region"),
+    )
+
+    sns_client.confirm_subscription(
+        TopicArn=data["TopicArn"],
+        Token=data["Token"],
+        AuthenticateOnUnsubscribe='true',
+    )
+
+    return Response()
+
+
+@view_config(
+    route_name="ses.hook",
+    require_methods=["POST"],
+    require_csrf=False,
+    header="x-amz-sns-message-type:Notification",
+)
+def notification(request):
```

Comment:
This route needs to lean towards failing hard to make Retry logic simple with SNS webhooks.

SNS webhook retry conditions:
- HTTP status in the range 500-599.
- HTTP status outside the range 200-599.

Looks like we should be OK, as the explicit `HTTPBadRequest` responses are indeed for requests that are successfully ignored, but something to keep in mind.